_source/_posts/2020-12-18-how-to-use-webauthn-csharp-dotnet.md

```diff
@@ -0,0 +1,780 @@
+
+Alternatively, you can install [Okta CLI](https://cli.okta.com/) and login to your Okta account with the `okta login` command or sign up for a new account with the `okta start` command if you don't already have an Okta account.
+
+Now, I will create a custom field for storing the user's public key. Go to **Users** in the menu, click **Profile Editor**, and click the **Profile** button next to the default Okta profile. Click **Add Attribute**, fill the `Add Attribute` form and click **Save**:
+
+{% img blog/webauthn-csharp/image8.png alt:"" width:"800" %}{: .center-image }
+
+Do the same to add another attribute called `CredentialId`.
+
+Next, go to **API** in the menu, click **Tokens**, and click the **Create Token** button. Enter the token name (I will use *OktaWebAuthn*) and click the **Create Token** button. Make sure you copy the token because you won't be able to view it again:
+
+{% img blog/webauthn-csharp/image9.png alt:"" width:"800" %}{: .center-image }
+
+Switch to Visual Studio, open the `appsettings.json` again and add a section for Okta:
+
+```json
+"Okta": {
+  "Domain": "https://{yourOktaDomain}",
+  "ApiToken": "{yourToken}"
+}
+```
+
+Finally, add a NuGet package for managing users in Okta:
+
+```sh
+dotnet add package Okta.Sdk
+```
+
+Open the `AccountController` class and add the `SaveCredentials` method:
+
+```cs
+public async Task<JsonResult> SaveCredentials([FromBody] AuthenticatorAttestationRawResponse attestationResponse)
+{
+    try
+    {
+        var jsonOptions = HttpContext.Session.GetString("fido2.attestationOptions");
+        var options = CredentialCreateOptions.FromJson(jsonOptions);
+
+        var fidoCredentials = await fido2.MakeNewCredentialAsync(attestationResponse, options, IsCredentialUnique);
+
+        var storedCredential = new StoredCredential{
+            Descriptor = new PublicKeyCredentialDescriptor(fidoCredentials.Result.CredentialId),
+            PublicKey = fidoCredentials.Result.PublicKey,
+            UserHandle = fidoCredentials.Result.User.Id,
+            SignatureCounter = fidoCredentials.Result.Counter,
+            CredType = fidoCredentials.Result.CredType,
+            RegDate = DateTime.Now,
+            AaGuid = fidoCredentials.Result.Aaguid
+        };
+
+        var result = await oktaClient.Users.CreateUserAsync(new CreateUserWithoutCredentialsOptions
+        {
+            Profile = new UserProfile
+            {
+                Login = options.User.Name,
+                Email = options.User.Name,
+                DisplayName = options.User.DisplayName,
+                ["CredentialId"] = Convert.ToBase64String(fidoCredentials.Result.CredentialId),
+                ["PasswordlessPublicKey"] = JsonConvert.SerializeObject(storedCredential)
+            }
+        });
+
+        return Json(fidoCredentials);
+    }
+    catch (Exception e)
+    {
+        return Json(new Fido2.CredentialMakeResult { Status = "error", ErrorMessage = e.Message });
+    }
+}
```

Comment:
Formatting (indent) is inconsistent. Above, there are 2-space indents for CS files. I try to be consistent by file type.